ldap/servers/slapd/back-ldbm/ldbm_search.c

```diff
@@ -999,13 +1005,33 @@ build_candidate_list(Slapi_PBlock *pb, backend *be, struct backentry *e, const c
         break;
 
     case LDAP_SCOPE_ONELEVEL:
-        *candidates = onelevel_candidates(pb, be, base, e, filter, managedsait,
-                                          lookup_returned_allidsp, &err);
+        /* Dup before optimise to prevent issues with aci and filter as intended vs optimised. */
+        filter_copy = slapi_filter_dup(filter);
+        /* modify the filter to be: (&(parentid=idofbase)(|(originalfilter)(objectclass=referral))) */
+        filter_copy = create_onelevel_filter(filter_copy, e, managedsait);
+        /* Now optimise the filter for use */
+        slapi_filter_optimise(filter_copy);
+
+        slapi_log_err(SLAPI_LOG_FILTER, "ldbm_back_search", "Optimised ONE filter to - %s\n",
+             slapi_filter_to_string(filter_copy, logbuf, sizeof(logbuf)));
+
+        *candidates = onelevel_candidates(pb, be, base, filter_copy, lookup_returned_allidsp, &err);
+        slapi_filter_free(filter_copy, 1);
         break;
 
     case LDAP_SCOPE_SUBTREE:
-        *candidates = subtree_candidates(pb, be, base, e, filter, managedsait,
-                                         lookup_returned_allidsp, &err);
+        /* Dup before optimise to prevent issues with aci and filter as intended vs optimised. */
+        filter_copy = slapi_filter_dup(filter);
```

Comment:
This filter duping & freeing for every search concerns me from a performance & memory frag perspective.  I wonder if create_*_filter() functions could be modified to only dup the filter if the filter is actually modified.  Return NULL if not modified, and use the original filter, otherwise use the copy.  That would at least reduce this overhead for cases where the optimization is not needed.

Reply:
But there aren't many cases where this would happen. Create onelevel always has to add the parentid element. Create subtree always has to add the objectClass referral (unless we get the managedsait control, which practically no one sends). There additions were what caused the issue before. 

About the only way to potential min-max this to avoid alloc/dealloc would be to pull out the optimised inner, but then we loose the ability to optimise around parentid= and oc=referral. 

Even with the suggestion to have a filter as intended vs as optimised, we have to do an extra dup/free. So that's unavoidable unless we do some weird tracking of the inner filter that we optimise, and then the wrapper that we don't, and we free them separately or something. Which risks more memory corruption and weirdness. Generally this all gets very messy ....

Reply:
Sure, makes sense.  I was just hoping it could be improved, but doesn't look that way.  Ok, thanks for listening :-)  This fix is going through the entire IPA test suite at the moment.  And hopefully this week I can do some testing numbers.  

I know @tbordaz had some concerns that this fix will impact a feature where we skip acl evaluation of filter components.  He thought this might actually break that - which in turn would hurt performance (or worse?).  I asked him to go into detail and comment on this PR.  He hasn't had a chance to do so yet, but I'd like his concern looked into (once they are accurately expressed).

In the meantime I hope to run some performance tests.  What filters do you suggest I test that will best showcase this feature's improvements?

Reply:
Wellllll after a walk I think there MIGHT be a way to do what you want ... so let me think a bit and come up with a new patch on top today.

The classic example is "(&(objectClass=person)(uid=foo))". But generally most of the filters that SSSD emits would benefit too, I remember one of the filters was pathologically bad, but I don't recall what it was now.


Since this is only a first pass, it'll only improve things like the above, but later we can make more interesting optimisations.